File: src/example_call_aerobulk.f90
! AeroBulk / 2015 / L. Brodeau

PROGRAM EXAMPLE_CALL_AEROBULK_COMPUTE

   USE mod_aerobulk
   USE mod_const

   IMPLICIT NONE

   INTEGER, PARAMETER :: nx = 2, &
      &                  ny = 1

   REAL(wp), DIMENSION(nx,ny) :: zsst, zt_zt, zq_zt, zU_zu, zV_zu, zslp, &
      &                         zRsw, zRlw,           &
      &                         zQL, zQH, zTau_x, zTau_y, zTs

   PRINT *, ''

   zsst = 22.
   PRINT *, ' *** SST = ', zsst, ' deg.C'
   zsst  = zsst + rt0  ! [K]

   zt_zt(1,1) = rt0 + 20.  ! [K]
   zt_zt(2,1) = rt0 + 25.  ! [K] ! second case is stable ABL as t_air > SST (25>22)!


   zq_zt = 0.012         ! [kg/kg]
   zU_zu = 4.            ! [m/s]
   zV_zu = 10.           ! [m/s]
   zslp  = 101000.       ! [Pa]

   zRsw  = 0. ! (night)  ! [W/m^2]
   zRlw  = 350.          ! [W/m^2]


   PRINT *, ''
   PRINT *, ' *********** COARE 3.6 *****************'
   CALL aerobulk_model( 'coare3p6', 2._wp, 10._wp, zsst, zt_zt, &
      &                 zq_zt, zU_zu, zV_zu, zslp,   &
      &                 zQL, zQH, zTau_x, zTau_y,    &
      &                 Niter=20, rad_sw=zRsw, rad_lw=zRlw, T_s=zTs )
   PRINT *, ''
   PRINT *, ' Wind speed at zu       =', REAL(SQRT(zU_zu*zU_zu + zV_zu*zV_zu),4), ' m/s'
   PRINT *, '    SST                 =', REAL(zsst,4), ' K'
   PRINT *, ' Air temperature at zt  =', REAL(zt_zt,4), ' K'  ; PRINT *, ''
   PRINT *, ' Sensible heat flux: QH =', REAL(zQH,4), ' W/m**2'
   PRINT *, '  Latent  heat flux: QL =', REAL(zQL,4), ' W/m**2'
   PRINT *, ' Skin temperature: SSST =', REAL(zTs-rt0,4), ' K'
   PRINT *, ' Tau_x =', REAL(zTau_x,4), ' N/m**2'
   PRINT *, ' Tau_y =', REAL(zTau_y,4), ' N/m**2'
   PRINT *, ' Tau   =', REAL(SQRT(zTau_x*zTau_x + zTau_y*zTau_y),4), ' N/m**2' ; PRINT *, ''
   PRINT *, ''

   PRINT *, ''
   PRINT *, ' *********** ECMWF *****************'
   CALL aerobulk_model( 'ecmwf', 2._wp, 10._wp, zsst, zt_zt, &
      &                 zq_zt, zU_zu, zV_zu, zslp,   &
      &                 zQL, zQH, zTau_x, zTau_y,    &
      &                 Niter=20, rad_sw=zRsw, rad_lw=zRlw, T_s=zTs )
   PRINT *, ''
   PRINT *, ' Wind speed at zu       =', REAL(SQRT(zU_zu*zU_zu + zV_zu*zV_zu),4), ' m/s'
   PRINT *, '    SST                 =', REAL(zsst,4), ' K' 
   PRINT *, ' Air temperature at zt  =', REAL(zt_zt,4), ' K' ; PRINT *, ''
   PRINT *, ' Sensible heat flux: QH =', REAL(zQH,4), ' W/m**2'
   PRINT *, '  Latent  heat flux: QL =', REAL(zQL,4), ' W/m**2'
   PRINT *, ' Skin temperature: SSST =', REAL(zTs-rt0,4), ' K'
   PRINT *, ' Tau_x =', REAL(zTau_x,4), ' N/m**2'
   PRINT *, ' Tau_y =', REAL(zTau_y,4), ' N/m**2'
   PRINT *, ' Tau   =', REAL(SQRT(zTau_x*zTau_x + zTau_y*zTau_y),4), ' N/m**2' ; PRINT *, ''
   PRINT *, ''

   PRINT *, ''
   PRINT *, ' *********** NCAR *****************'
   CALL aerobulk_model( 'ncar', 2._wp, 10._wp, zsst, zt_zt, &
      &                 zq_zt, zU_zu, zV_zu, zslp,   &
      &                 zQL, zQH, zTau_x, zTau_y,    &
      &                 Niter=20 )
   PRINT *, ''
   PRINT *, ' Wind speed at zu       =', REAL(SQRT(zU_zu*zU_zu + zV_zu*zV_zu),4), ' m/s'
   PRINT *, '    SST                 =', REAL(zsst,4), 'K'
   PRINT *, ' Air temperature at zt  =', REAL(zt_zt,4), 'K' ; PRINT *, ''
   PRINT *, ' Sensible heat flux: QH =', REAL(zQH,4), ' W/m**2'
   PRINT *, '  Latent  heat flux: QL =', REAL(zQL,4), ' W/m**2'
   PRINT *, ' Tau_x =', REAL(zTau_x,4), ' N/m**2'
   PRINT *, ' Tau_y =', REAL(zTau_y,4), ' N/m**2'
   PRINT *, ' Tau   =', REAL(SQRT(zTau_x*zTau_x + zTau_y*zTau_y),4), ' N/m**2' ; PRINT *, ''
   PRINT *, ''

END PROGRAM EXAMPLE_CALL_AEROBULK_COMPUTE

File: src/example_call_aerobulk.cpp
/* -*- mode: c++; coding: utf-8; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4; show-trailing-whitespace: t -*- vim: set fenc=utf-8 ft=cpp et sw=4 ts=4 sts=4: */

#include "aerobulk.hpp"
#include <iostream>

int main(int argc, char** argv)
{

    double zt =  2.; // height of measurement for air temperature and humidity [m]
    double zu = 10.; // height of measurement for wind speed [m]

    std::vector<double> zsst  = {273.15 + 22., 273.15 + 22.}; // sea surface temperature [K]
    std::vector<double> zt_zt = {273.15 + 20., 273.15 + 25.}; // air absolute temperature at zt [K] ## second case is stable ABL as t_air > SST (25>22)!
    std::vector<double> zq_zt = {0.012, 0.012};               // air specific humidity at zt [g/kg]
    std::vector<double> zU_zu = { 4.,  4.};                   // u-wind speed component at zu [m/s]
    std::vector<double> zV_zu = {10., 10.};                   // v-wind speed component at zu [m/s]
    std::vector<double> zslp  = {101000.0, 101000.0};         // sea-level atmospheric pressure [Pa]

    std::vector<double> zRsw  = {0., 0.};       // downwelling shortwave (solar)     radiation [W/m^2] ## night!
    std::vector<double> zRlw  = {350., 350.};   // downwelling longwave  (infra-red) radiation [W/m^2] 

    // AeroBulk output:
    std::vector<double> zQL;
    std::vector<double> zQH;
    std::vector<double> zTau_x;
    std::vector<double> zTau_y;
    std::vector<double> zT_s;

    aerobulk::model(aerobulk::algorithm::COARE3p6, zt, zu, zsst, zt_zt,
                    zq_zt, zU_zu, zV_zu, zslp,
                    zQL, zQH, zTau_x, zTau_y,
                    zRsw, zRlw, zT_s );

    std::cout
        << "\n *********** COARE 3.6 *****************\n"
        << " QH = \t" << zQH[0] << "\t" << zQH[1] << std::endl
        << " QL = \t" << zQL[0] << "\t" << zQL[1] << std::endl
        << " Tau_x = \t" << zTau_x[0] << "\t" << zTau_x[1] << std::endl
        << " Tau_y = \t" << zTau_y[0] << "\t" << zTau_y[1] << std::endl
        << " T_s = \t" << zT_s[0] << "\t" << zT_s[1] << std::endl;

    std::vector<double> L = aerobulk::l_vap(zsst);

    std::cout << " Evaporation = \t" << -zQL[0]*1e-3/L[0] << "\t" << -zQL[1]*1e-3/L[1] << std::endl;


    aerobulk::model(aerobulk::algorithm::ECMWF, zt, zu, zsst, zt_zt,
                    zq_zt, zU_zu, zV_zu, zslp,
                    zQL, zQH, zTau_x, zTau_y,
                    zRsw, zRlw, zT_s );

    std::cout
        << "\n *********** ECMWF *****************\n"
        << " QH = \t" << zQH[0] << "\t" << zQH[1] << std::endl
        << " QL = \t" << zQL[0] << "\t" << zQL[1] << std::endl
        << " Tau_x = \t" << zTau_x[0] << "\t" << zTau_x[1] << std::endl
        << " Tau_y = \t" << zTau_y[0] << "\t" << zTau_y[1] << std::endl
        << " T_s = \t" << zT_s[0] << "\t" << zT_s[1] << std::endl;

    L = aerobulk::l_vap(zsst);

    std::cout << " Evaporation = \t" << -zQL[0]*1e-3/L[0] << "\t" << -zQL[1]*1e-3/L[1] << std::endl;

    aerobulk::model(aerobulk::algorithm::NCAR, 2, 10, zsst, zt_zt,
            zq_zt, zU_zu, zV_zu, zslp,
            zQL, zQH, zTau_x, zTau_y);

    std::cout
        << "\n *********** NCAR *****************\n"
        << " QH = \t" << zQH[0] << "\t" << zQH[1] << std::endl
        << " QL = \t" << zQL[0] << "\t" << zQL[1] << std::endl
        << " Tau_x = \t" << zTau_x[0] << "\t" << zTau_x[1] << std::endl
        << " Tau_y = \t" << zTau_y[0] << "\t" << zTau_y[1] << std::endl
        << " T_s = \t" << zsst[0] << "\t" << zsst[1] << std::endl;

    L = aerobulk::l_vap(zsst);

    std::cout << " Evaporation = \t" << -zQL[0]*1e-3/L[0] << "\t" << -zQL[1]*1e-3/L[1] << std::endl;
}
 

